File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/api/Robot.java
package org.firstinspires.ftc.teamcode.api;

import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode;

public class Robot {
    private final LinearOpMode opMode;

    public static final double motor1Angle = 0.0;
    public static final double motor2Angle = Math.PI * (2.0 / 3.0);
    public static final double motor3Angle = 2.0 * Math.PI * (2.0 / 3.0);

    public Robot(LinearOpMode opMode) {
        this.opMode = opMode;
    }

    /**
     * Returns the power of each motor from a joystick's polar coordinates.
     *
     * @param radians The angle of the joystick.
     * @param magnitude The distance from center of the joystick.
     * @return The power of each wheel.
     */
    public static TripleTuple motorPower(double radians, double magnitude) {
        return new TripleTuple(
                magnitude * Math.sin(Robot.motor1Angle - radians),
                magnitude * Math.sin(Robot.motor2Angle - radians),
                magnitude * Math.sin(Robot.motor3Angle - radians)
        );
    }

    /**
     * Returns the power of each motor from a joystick coordinate.
     *
     * @param x The x-position of the joystick.
     * @param y The y-position of the joystick.
     * @return The power of each wheel.
     */
    public static TripleTuple motorPowerXY(double x, double y) {
        return motorPower(
                Math.atan2(y, x) - (Math.PI / 2.0),
                Math.sqrt(y * y + x * x)
        );
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/api/arch/Robot.kt
package org.firstinspires.ftc.teamcode.api.arch

import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode
import org.firstinspires.ftc.teamcode.api.arch.runtime.RuntimeBuilder

const val DEFAULT_ORDER: Byte = 0

open class Robot(val teleop: LinearOpMode, val cfg: Config) {
    private val runtimeBuilder: RuntimeBuilder = RuntimeBuilder()
    private val ctx: Context = Context(teleop, cfg)

    init {
        this.configure(this.runtimeBuilder)
    }

    open fun configure(builder: RuntimeBuilder) {}

    fun run() {
        val runtime = runtimeBuilder.build(this.cfg)
        runtime.run(this.ctx)
    }
}
